File: lib/src/main/java/uk/iatom/rfm69/registers/values/FifoThresh.java
package uk.iatom.rfm69.registers.values;

import java.util.Arrays;
import java.util.stream.Collectors;

import uk.iatom.rfm69.registers.IRegisterValue;
import uk.iatom.rfm69.registers.SingleRegister;

/**
 * Binding for RegFifoThresh
 */
public enum FifoThresh implements IRegisterValue {
    TXSTART_FIFOTHRESH(0x00), // Reset value
    TXSTART_FIFONOTEMPTY(0x80), // Recommended default
    DEFAULT(0x0F); // Default

    public static FifoThresh[] inverse = Arrays.stream(values())
            .collect(Collectors.groupingBy(f -> f.val, Collectors.reducing(null, (a, b) -> b)))
            .values().stream()
            .toArray(FifoThresh[]::new);

    private byte val;

    FifoThresh(int val) {
        this.val = (byte) val;
    }

    @Override
    public byte val() {
        return val;
    }

    public static Byte encode(FifoThresh fifo) {
        return fifo.val;
    }

    public static FifoThresh decode(Byte b) {
        return SingleRegister.helpDecodeEnum(values(), b);
    }
}

File: lib/src/main/java/uk/iatom/rfm69/registers/values/FifoThresh.kt
package uk.iatom.rfm69.registers.values

import uk.iatom.rfm69.registers.IRegisterValue
import uk.iatom.rfm69.registers.SingleRegister.Companion.helpDecodeEnum
import java.util.*
import java.util.function.*
import java.util.stream.*

/**
 * Binding for RegFifoThresh
 */
enum class FifoThresh(`val`: Int): IRegisterValue {


    TXSTART_FIFOTHRESH(0x00),


    // Reset value
    TXSTART_FIFONOTEMPTY(0x80),


    // Recommended default
    DEFAULT(0x0F);

    private val `val`: Byte

    init {
        this.`val` = `val`.toByte()
    }

    override fun `val`(): Byte {
        return `val`
    }

    companion object {


        // Default
        var inverse = Arrays
                .stream<FifoThresh?>(entries.toTypedArray())
                .collect(
                        Collectors.groupingBy<FifoThresh?, Byte, Any, FifoThresh?>(
                                Function<FifoThresh?, Byte> { f: FifoThresh? -> f!!.`val` },
                                Collectors.reducing<FifoThresh?>(
                                        null,
                                        BinaryOperator<FifoThresh?> { a: FifoThresh?, b: FifoThresh? -> b })
                                                                                  )
                        ).values
                .stream()
                .toArray<FifoThresh> { _Dummy_.__Array__() }

        fun encode(fifo: FifoThresh): Byte {
            return fifo.`val`
        }

        fun decode(b: Byte?): FifoThresh {
            return helpDecodeEnum(
                    entries.toTypedArray(),
                    b!!
                                 )
        }
    }
}
